### PR TITLE
wallet defragmentation

### PR DESCRIPTION
**Pull request by johnathanhowell**
This PR introduces wallet defragmentation, in order to solve the problems noticed by some users, notably Poloniex, where wallets with many outputs would fail to send coins with `transaction too large`. This PR adds `defragWallet()`, which is called after every block and sends the sum of the 15 largest outputs to itself.

### CODE REVIEW

File: modules/wallet/defrag.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	var so sortedOutputs
+	for scoid, sco := range w.siacoinOutputs {
+		so.ids = append(so.ids, scoid)
+		so.outputs = append(so.outputs, sco)
+	}
+	sort.Sort(sort.Reverse(so))
+
+	defragOutputs := so.outputs[:15]
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: 15 should be constant

File: modules/wallet/defrag.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// defragWallet computes the sum of the 15 largest outputs in the wallet and
+// sends that sum to itself, effectively defragmenting the wallet. This defrag
+// operation is only performed if the wallet has greater than defragThreshold
+// outputs.
+func (w *Wallet) defragWallet() {
+	if len(w.siacoinOutputs) < defragThreshold {
+		return
+	}
+
+	var so sortedOutputs
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: comments

File: modules/wallet/defrag.go (hunk 1 above)
Comment (DavidVorick) [inline modules/wallet/defrag.go]: also comments

File: modules/wallet/defrag.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	w.log.Printf("defragmenting wallet: %v outputs, %v total value\n", len(defragOutputs), totalOutputValue)
+
+	var dest types.UnlockHash
+	for h := range w.keys {
+		dest = h
+		break
+	}
+
+	go func() {
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: comments. Why is this in a gofunc?

File: modules/wallet/defrag.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	// defragThreshold is the number of outputs a wallet is allowed before it is
+	// defragmented.
+	defragThreshold = 20
+)
+
+// defragWallet computes the sum of the 15 largest outputs in the wallet and
+// sends that sum to itself, effectively defragmenting the wallet. This defrag
+// operation is only performed if the wallet has greater than defragThreshold
+// outputs.
+func (w *Wallet) defragWallet() {
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: There should probably be some check that the previous defrag operation succeeded / got into a block before making another one.

Comment (lukechampine) [inline modules/wallet/defrag.go]: In the contractor I use a `TryMutex` to accomplish this: https://github.com/NebulousLabs/Sia/blob/master/modules/renter/contractor/update.go#L66
Not sure if it's the best pattern but it seems to work.

Comment (johnathanhowell) [inline modules/wallet/defrag.go]: I think using a trylock would only prevent multiple defrags from running concurrently, it wouldn't tell you if the previous defrag transaction was accepted into the block

Comment (DavidVorick) [inline modules/wallet/defrag.go]: the error is a graceful/benign one, so we decided check not needed

File: modules/wallet/defrag.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+		w.log.Println("error getting an address for defrag: ", err)
+		return
+	}
+
+	// send the sum of the outputs to this wallet. This operation is done in a
+	// goroutine since defragWallet() is called under lock.
+	go func() {
+		txns, err := w.SendSiacoins(totalOutputValue, addr.UnlockHash())
+		if err != nil {
+			w.log.Printf("error sending defrag transaction with value %v: %v\n ", totalOutputValue, err)
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: These log statements should have full punctuation, they are being displayed to end users!

Comment (DavidVorick) [inline modules/wallet/defrag.go]: "Attempted to defragement wallet but failed. Defragment size: %v. Error: %v.", totalOutputValue, err

File: modules/wallet/defrag.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	sort.Sort(sort.Reverse(so))
+
+	// choose the defragBatchSizeth largest outputs from the wallet and sum them
+	defragOutputs := so.outputs[:defragBatchSize]
+	var totalOutputValue types.Currency
+	for _, output := range defragOutputs {
+		totalOutputValue = totalOutputValue.Add(output.Value)
+	}
+
+	w.log.Printf("performing wallet defragmentation transaction: %v outputs, %v total value\n", len(defragOutputs), totalOutputValue)
```
Comment (DavidVorick) [inline modules/wallet/defrag.go]: Needs full punctuation

Comment (DavidVorick) [inline modules/wallet/defrag.go]: But, really this log should be merged with the one below, no need to make two logging statements per defrag, just one is fine (one if success, a different one for each type of error (I think only 1 type possible right now))

**Comment by johnathanhowell** [comment]
ok, I've made the defrag only print one log per cycle, and made the logs more human-readable and informative as requested.

**Pull request merged**